### ISSUE TITLE
pymysql Connection.ping AttributeError

### ISSUE THREAD

**Issue by luoye2018** [sqlalchemy/sqlalchemy#10650]
### Describe the bug

sqlalchemy with PyMySQL 0.9.3 ping cause exception.

### SQLAlchemy Version in Use

2.0.23 1.4.50

### DBAPI (i.e. the database driver)

pymysql

### Database Vendor and Major Version

MySQL 8

### Python Version

3.9

### Operating system

Linux

### To Reproduce

```python
from sqlalchemy import create_engine
from sqlalchemy import text

SQLALCHEMY_DATABASE_URI='mysql+pymysql://xxxx:xxxx@xxxx:3306/mysql'

engine = create_engine(SQLALCHEMY_DATABASE_URI, pool_pre_ping=True, pool_size=1)

def test():
    with engine.connect() as conn:
        conn.execute(text("select user from mysql.user limit 1"))

for _ in range(2):
    test()
```

### Error

```
Traceback (most recent call last):
  File "/workspace/test-env/../greatrds/.vscode/test.py", line 13, in <module>
    test()
  File "/workspace/test-env/../greatrds/.vscode/test.py", line 9, in test
    with engine.connect() as conn:
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3268, in connect
    return self._connection_cls(self)
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 3292, in raw_connection
    return self.pool.connect()
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1378, in _checkout
    del fairy
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 146, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1306, in _checkout
    result = pool._dialect._do_ping_w_event(
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 709, in _do_ping_w_event
    return self.do_ping(dbapi_connection)
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/pymysql.py", line 104, in do_ping
    if self._send_false_to_ping:
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 1146, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/workspace/test-env/venv/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/pymysql.py", line 93, in _send_false_to_ping
    insp = langhelpers.get_callable_argspec(Connection.ping)
AttributeError: 'function' object has no attribute 'ping'
```

### Additional context

Package    Version
---------- -------
greenlet   3.0.1
pip        21.2.4
PyMySQL    0.9.3
setuptools 58.1.0
SQLAlchemy 1.4.50

**Comment by zzzeek** [comment]
hi there -

pymysql 0.9.3 is five years old.  any reason you can't simply upgrade to a modern version of the library?

**Comment by CaselIT** [comment]
Let's place a raise requiring a min version in the library instead

**Comment by zzzeek** [comment]
might be a good idea.  still, would like to know if some kind of conda / linux /something is forcing people to use pymysql 0.9.3 while giving them SQLAlchemy 1.4.  SQLAlchemy 1.4 goes back three years.

**Comment by CaselIT** [comment]
Sqlalchemy seems the latest one, 2.0.23

**Comment by luoye2018** [comment]
> might be a good idea. still, would like to know if some kind of conda / linux /something is forcing people to use pymysql 0.9.3 while giving them SQLAlchemy 1.4. SQLAlchemy 1.4 goes back three years.

Sqlalchemy 2.0.23 and 1.4.50 has same promblem，1.4.50 introduce problems Released: Oct 30, 2023
this promblem resolve on pymysql 1.0.0  Released: Jan 7, 2021

```
def Connect(*args, **kwargs):
    """
    Connect to the database; see connections.Connection.__init__() for
    more information.
    """
...

connect = Connection = Connect
```
```
        try:
            Connection = __import__("pymysql.connections").Connection
        except (ImportError, AttributeError):
            return True
        else:
            insp = langhelpers.get_callable_argspec(Connection.ping)

```
Connection get function but not class .
my fix is add check ping,
maybe we raise version check
```
        try:
            Connection = __import__("pymysql.connections").Connection
            getattr(Connection, "ping")
        except (ImportError, AttributeError):
            return True
```

__import__  function will return the top level module of a package.
__import__("pymysql.connections") only import pymysql
if we need pymysql.connections module we need 
```
 __import__('pymysql.connections',fromlist=[object])
```
or 
```
importlib.import_module('pymysql.connections')
```

https://docs.python.org/3/library/functions.html#import__

**Comment by luoye2018** [comment]
> Sqlalchemy seems the latest one, 2.0.23

pymysql 1.0.0 will resolve this problem, maybe we can change import function

**Comment by CaselIT** [comment]
@zzzeek while I think for v2 it's fine to add a lower bound to the dialect I'm not sure for 1.4 it's doable, so maybe detecting it would be a better option?

**Comment by zzzeek** [comment]
we need to add a lower bound for 2.0.  for 1.4, this is not a regression, has an easy end-user fix, and we're well on our way to EOLing 1.4 soon so I would not argue for a backport

**Comment by CaselIT** [comment]
I kinda thing there's a bug there
we should have done
```py
Connection = __import__("pymysql.connections").connections.Connection
```
instead of 
```py
Connection = __import__("pymysql.connections").Connection
```

**Comment by zzzeek** [comment]
I would vote for

```py
>>> __import__("pymysql").Connection
<class 'pymysql.connections.Connection'>

```

that's not related to this issue though

**Comment by zzzeek** [comment]
since the code is already being extrmely defensive we can make it also defend against "ping" not present

but no 1.4 backport

**Comment by CaselIT** [comment]
but on pymysql<1 you get
```py
>>> __import__("pymysql").Connection
<function Connect at 0x7fab6b1f11c0>
```

that's where the issue comes from
https://github.com/PyMySQL/PyMySQL/blob/37fd1e1b0126d75d80eef59c053f80634b09bd75/pymysql/__init__.py#L88-L108

**Comment by zzzeek** [comment]
oh, it has Ping still?   fine, fix for 2.0.

**Comment by CaselIT** [comment]
no pymysql did stupid stuff in v0.x aliasing connection with a function for some strange reason

**Comment by zzzeek** [comment]
this can be a 1.4 backport assuming 1.4.49 did not have this problem.   It looked like the issue was old pymysql didn't have "ping" at all, but if it's local to the 1.4.50 change, then we can backport, but it might not be released

**Comment by CaselIT** [comment]
> this can be a 1.4 backport assuming 1.4.49 did not have this problem.

test seem to pass with 1.4.49 and PyMySQL<1

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Fix pre-ping regression in old PyMySQL versions.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4980

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **rel_2_0** branch:

**Fix pre-ping regression in old PyMySQL versions.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4981

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **rel_1_4** branch:

**Fix pre-ping regression in old PyMySQL versions.** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4982

**Comment by luoye2018** [comment]
> but on pymysql<1 you get
> ```py
> >>> __import__("pymysql").Connection
> <function Connect at 0x7fab6b1f11c0>
> ```
> 
> that's where the issue comes from
> https://github.com/PyMySQL/PyMySQL/blob/37fd1e1b0126d75d80eef59c053f80634b09bd75/pymysql/__init__.py#L88-L108

try it
```
__import__('pymysql.connections',fromlist=[object]).Connection
```

**Comment by luoye2018** [comment]
> this can be a 1.4 backport assuming 1.4.49 did not have this problem.   It looked like the issue was old pymysql didn't have "ping" at all, but if it's local to the 1.4.50 change, then we can backport, but it might not be released

change import it will be right

__import__('pymysql.connections',fromlist=[object])

**Comment by CaselIT** [comment]
It's now merged in the 1.4 branch.
As mike mentioned it man not get released on pypi, since 1.4 is so far along that any new release risk adding more bug than it solves, as evidenced by this exact issue :)

Feel free to install from the branch directly through
